providers/providers.go

```diff
@@ -92,9 +93,15 @@ func (pm *ProviderManager) providersForKey(k []byte) ([]peer.ID, error) {
 }
 
 func (pm *ProviderManager) getProvSet(k []byte) (*providerSet, error) {
-	cached, ok := pm.providers.Get(string(k))
-	if ok {
-		return cached.(*providerSet), nil
+	// If fetching all Provider Records store, purge cache so that
+	// we pull the complete list
+	if k == nil {
+		pm.cache.Purge()
```

Comment:
The way it currently works is that if there is a value in the cache, it doesn't touch the datastore. So if you have the information of 3 providers in the datastore, but the cache only has 2, then when you list you only see 2 and given that the goal of listing them all is to see "everything we have", purging the cache gets us that.

Alternatively, now that I think about it, I can just jump over the cache check rather than purging it. What do you think?

Reply:
Gotcha, one more question then, why would the cache have out of date values in it? That sounds like a 🐛...

Either it's the most up to date or missing right? Or is there a reach around? If so then yeah I guess avoid the cache altogether.

Reply:
OH YOU ARE RIGHT! :D

Reply:
See https://github.com/libp2p/go-libp2p-kad-dht/pull/490/commits/0947139776a3d988f5c23720d7f7973b90995cd1